rfc/001-composable-pipelines.md

```diff
@@ -0,0 +1,292 @@
+data. Otherwise data races may occur. Consequently, we need transactional
+interface (ACID) to support this operation.
+
+As long as VAST owns the underlying data (i.e., only VAST is allowed to make
+changes), VAST can already mutate of data at rest. [Spatial
+compaction][mutate-at-rest] uses a pre-defined disk quota as trigger to apply
+pipelines to a subset of to-be-transformed data. After VAST applies the
+pipeline, VAST optionally removes the original data in an atomic fashion.
+
+[mutate-at-rest]: https://vast.io/docs/use-vast/transform#modify-data-at-rest
+
+Modeled after the `compaction` plugin, we may consider exposing a mutable
+pipeline interface through a dedicate command. This command interacts with a
+node, and could be seen as a combination of `pull` and `push` that executes
+remotely. For example:
+
+```bash
+vast mutate --types=A,B,C 'where 10.0.0.0/8 | put orig.h resp.h'
+```
+
+This would apply the pipeline to filter and project in schema types `A`, `B`,
+and `C`.
+
+## Implementation
+
+We could implement the entire pipeline with Arrow IPC streams in the "interior"
+operators, with only `from` and `to` taking raw input as byte stream. But this
+may be too restrictive. Let's assume that input and output are dynamically
+typed. An operator must then define its input and output types, e.g., `from<In,
+Out>` takes as input `In` and produces instances of `Out`.
+
+### Matcher Example
+
+This can come in handy for working with the matcher plugin. Consider the use
+case of constructing a matcher from data. Consider this pipeline:
+
+```bash
+vast exec 'read csv |
+           matcher build --extract=net.src.ip' > ips.flatbuffer
+```
+
+Here, we have:
+
+- `read<Bytes, Arrow>`: reads `Bytes` and produces `Arrow` record batches
+- `matcher build<Arrow, Flatbuffer>`: consumes `Arrow` and create a matcher
+  backend.
+
+Let `Void` be a valid type where the operator as only a side effect. Then we can
+pre-load a matcher at a VAST node as follows:
+
+```bash
+vast exec 'from file:///ips.flatbuffer |
+           matcher load --name=ips vast:///1.2.3.4'
+```
+
+Or more UNIX'ish, assuming an implicit `from -`:
+
+```bash
+vast exec 'matcher load --name=ips' < ips.flatbuffer
+```
+
+Here, `matcher load<Flatbuffer, Void>` is the operator that updates the remote
+matcher state.
+
+Finally, the `matcher` sub-command `match` does the actual matching:
+
+```bash
+vast pull 'where #type == "suricata.flow" |
+           matcher match --on=:addr ips'
+```
+
+### Pipeline Management
+
+The above is a locally running pipeline, but most likely we want this to run at
+a node remotely, independent of where the client is. This begs the challenge:
+how do we manage all these pipelines? As with all other VAST functions, there
+could be a client command.
+
+```bash
+# Show all pipelines:
+vast pipeline list
+```
+
+Create remote pipelines at ease:
+
+```bash
+vast pipeline create foo 'where #type == "suricata.flow" | matcher match ips'
```

Comment:
what does it mean to have this as a remote pipeline? Does the implicit `from -` implicitely become `all data pushed to the server`? That might be a bit too much "impliciteness" in my opinion 😄, but it would work. The serverside dual of `from -` would be something like:
- `pushed<Void,Arrow><Remote>`

Reply:
I was thinking that we could have a dummy operator to indicate "the whole DB", "the input stream", or "both". For example:

- `new | where #type ~ /zeek.*/` to filter all Zeek logs on the ingest stream
- `old | where dest_ip in 192.168.0.64/27` to emulate a historical query
- `all | ...` to fuse the feed of `old` and `new` (aka. "unified query")

Names of course TBD. 😅

Reply:
I'm resolving this now. The topic of remote pipelines has its own RFC: #2577